<a href="https://colab.research.google.com/github/skyqi/LangChain/blob/main/ConversationSummaryBufferMemory%5B%E4%B8%80%E7%A7%8D%E6%B7%B7%E5%90%88%E8%AE%B0%E5%BF%86%E6%A8%A1%E5%9E%8B%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ConversationSummaryBufferMemory一种混合记忆模型
  ConversationSummaryBufferMemory，即对话总结缓冲记忆，它是一种混合记忆模型，结合了上述各种记忆机制，包括 ConversationSummaryMemory和ConversationBufferWindowMemory的特点。


当最新的对话文字长度在token 300字之内的时候，LangChain 会记忆原始对话内容；当对话文字超出了这个参数的长度，那么模型就会把所有超过预设长度的内容进行总结，以节省 Token 数量。

用于长时间聊天

In [1]:
import os
import getpass

api_key = os.getenv("ZHIPUAI_API_KEY")
if api_key is None:
  os.environ["ZHIPUAI_API_KEY"] = getpass.getpass()

··········


In [2]:
!pip install --upgrade --quiet  langchain-google-genai pillow
!pip install langchain langchain_community
!pip install --upgrade httpx httpx-sse PyJWT

from langchain_community.chat_models import ChatZhipuAI
from langchain_community.chat_message_histories import ChatMessageHistory
from langchain_core.runnables.history import RunnableWithMessageHistory
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from langchain.chains import ConversationChain

llm = ChatZhipuAI(
    model="glm-4",
    temperature=0.5,
)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 40.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 314.7/314.7 kB 33.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 125.2/125.2 kB 16.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.0/53.0 kB 6.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 142.7/142.7 kB 18.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
imageio 2.31.6 requires pillow<10.1.0,>=8.3.2, but you have pillow 10.3.0 which is incompatible.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 974.0/974.0 kB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 11.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.2/49.2 kB 7.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.6/75.6 kB 2.0 MB/s eta 0:00:00

In [3]:
#from langchain.memory import ConversationSummaryMemory
from langchain.chains.conversation.memory import ConversationSummaryBufferMemory

llm = ChatZhipuAI(
    model="glm-4",
    temperature=0.5,
)

memory = ConversationSummaryBufferMemory(llm=llm,max_token_limit=100)

In [4]:
conversation_with_summary = ConversationChain(
    llm = llm,
    memory= memory,
    verbose=True
)

In [5]:
conversation_with_summary.predict(input="如何保持健康的生活方式")



> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:

Human: 如何保持健康的生活方式
AI:


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]


> Finished chain.


'要保持健康的生活方式，可以从以下几个方面入手：\n\n1. 均衡饮食：遵循“膳食指南”，保证摄入充足的蔬菜、水果、全谷物和优质蛋白质，限制高糖、高盐和高脂肪的食物摄入。\n\n2. 适量运动：每周至少进行150分钟的中等强度有氧运动，如快走、游泳或骑自行车，也可以进行力量训练，增强肌肉和骨骼。\n\n3. 充足睡眠：成年人每天应保证7-8小时的高质量睡眠，保持良好的作息规律。\n\n4. 心理健康：保持积极的心态，学会缓解压力，与家人、朋友保持良好的社交关系。\n\n5. 定期体检：根据年龄和健康状况，定期进行体检，了解自己的健康状况，早期发现并治疗疾病。\n\n6. 健康生活方式：不吸烟，限制饮酒，避免长时间熬夜。\n\n7. 终身学习：保持好奇心，不断学习新知识，有助于保持大脑活力，延缓衰老。\n\n遵循这些原则，可以帮助你保持健康的生活方式。当然，每个人的具体情况不同，需要根据自己的实际情况进行调整。你还有其他问题吗？我会尽力为你提供帮助。'

要保持健康的生活方式，可以从以下几个方面入手：

1. 均衡饮食：遵循“膳食指南”，保证摄入充足的蔬菜、水果、全谷物和优质蛋白质，限制高糖、高盐和高脂肪的食物摄入。

2. 适量运动：每周至少进行150分钟的中等强度有氧运动，如快走、游泳或骑自行车，也可以进行力量训练，增强肌肉和骨骼。

3. 充足睡眠：成年人每天应保证7-8小时的高质量睡眠，保持良好的作息规律。

4. 心理健康：保持积极的心态，学会缓解压力，与家人、朋友保持良好的社交关系。

5. 定期体检：根据年龄和健康状况，定期进行体检，了解自己的健康状况，早期发现并治疗疾病。

6. 健康生活方式：不吸烟，限制饮酒，避免长时间熬夜。

7. 终身学习：保持好奇心，不断学习新知识，有助于保持大脑活力，延缓衰老。

遵循这些原则，可以帮助你保持健康的生活方式。当然，每个人的具体情况不同，需要根据自己的实际情况进行调整。你还有其他问题吗？我会尽力为你提供帮助。

In [6]:
conversation_with_summary.predict(input="有效的学习方法和技巧")



> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
System: The human inquires about how to maintain a healthy lifestyle. The AI suggests several key aspects for maintaining health, including a balanced diet, moderate exercise, sufficient sleep, mental health care, regular health check-ups, adoption of healthy living habits, and lifelong learning. The AI offers to assist with any further questions. 

New summary: The discussion revolves around the human's question on how to maintain a healthy lifestyle, with the AI providing a comprehensive list of recommendations including balanced eating, regular physical activity, adequate sleep, psychological well-being, regular health check-ups, healthy living practices, and contin

'AI: 有效的学习方法和技巧对于维持健康的生活方式同样重要。以下是一些建议：\n\n1. **目标设定**：明确你的学习目标，它们应该是具体、可衡量、可达到、相关性强和时限性的（SMART原则）。\n2. **规划时间**：合理安排学习时间，确保有规律的学习和复习。\n3. **主动学习**：不仅仅是被动接受信息，而是要主动提问，参与讨论，进行实践。\n4. **多样化学习**：运用不同的学习方式和技巧，比如视觉、听觉和动手操作，以增强记忆和理解。\n5. **分块学习**：将复杂的信息分成小块，逐一学习，这样更容易消化和记忆。\n6. **定期复习**：使用间隔重复技术，定期回顾旧信息，以强化记忆。\n7. **充足休息**：保证有足够的睡眠，这对于记忆和学习至关重要。\n8. **保持好奇心**：对所学内容保持兴趣和好奇心，这可以提高学习的动力和效率。\n9. **合理饮食**：健康的饮食习惯也能促进大脑功能的最佳状态。\n10. **锻炼身体**：适量的身体活动可以提高心血管健康，进而促进大脑健康。\n\n如果你有特定的学习领域或需要针对特定的学科技巧，我可以提供更具体的建议。你希望在哪方面提高学习效率呢？'

AI: 有效的学习方法和技巧对于维持健康的生活方式同样重要。以下是一些建议：

1. **目标设定**：明确你的学习目标，它们应该是具体、可衡量、可达到、相关性强和时限性的（SMART原则）。
2. **规划时间**：合理安排学习时间，确保有规律的学习和复习。
3. **主动学习**：不仅仅是被动接受信息，而是要主动提问，参与讨论，进行实践。
4. **多样化学习**：运用不同的学习方式和技巧，比如视觉、听觉和动手操作，以增强记忆和理解。
5. **分块学习**：将复杂的信息分成小块，逐一学习，这样更容易消化和记忆。
6. **定期复习**：使用间隔重复技术，定期回顾旧信息，以强化记忆。
7. **充足休息**：保证有足够的睡眠，这对于记忆和学习至关重要。
8. **保持好奇心**：对所学内容保持兴趣和好奇心，这可以提高学习的动力和效率。
9. **合理饮食**：健康的饮食习惯也能促进大脑功能的最佳状态。
10. **锻炼身体**：适量的身体活动可以提高心血管健康，进而促进大脑健康。

如果你有特定的学习领域或需要针对特定的学科技巧，我可以提供更具体的建议。你希望在哪方面提高学习效率呢？

注意：AI的回复跟上次是总结性的


In [10]:
history = conversation_with_summary.memory.chat_memory.messages
history

[]

In [12]:
memory.load_memory_variables({})

{'history': 'System: The conversation begins with the human seeking advice on maintaining a healthy lifestyle, to which the AI offers a detailed list of recommendations. This includes a balanced diet, exercise, sufficient sleep, mental health care, and other habits. The discussion then shifts to the human inquiring about effective learning methods and techniques, with the AI providing a comprehensive guide that encompasses setting clear goals, time management, active learning, varied approaches, chunking information, regular review, adequate rest, curiosity, healthy eating, and physical activity. The AI also extends the offer for more tailored advice based on specific learning areas or subjects.'}

{'history': '系统：对话从人类寻求保持健康生活方式的建议开始，人工智能会提供详细的建议清单。这包括均衡饮食、锻炼、充足睡眠、心理健康护理和其他习惯。然后讨论转向人类询问有效的学习方法和技巧，人工智能会提供全面的指南，包括设定明确的目标、时间管理、主动学习、多种方法、分块信息、定期复习、充分休息、好奇心、健康饮食和身体活动。人工智能还会根据特定的学习领域或科目提供更有针对性的建议。'}

In [13]:
conversation_with_summary.predict(input="我姐姐明天要过生日，我需要一束生日花束。")



> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
System: The conversation begins with the human seeking advice on maintaining a healthy lifestyle, to which the AI offers a detailed list of recommendations. This includes a balanced diet, exercise, sufficient sleep, mental health care, and other habits. The discussion then shifts to the human inquiring about effective learning methods and techniques, with the AI providing a comprehensive guide that encompasses setting clear goals, time management, active learning, varied approaches, chunking information, regular review, adequate rest, curiosity, healthy eating, and physical activity. The AI also extends the offer for more tailored advice based on specific learning area

'AI: 生日花束的选择取决于你姐姐的个人喜好，比如她喜欢的颜色、花种和香味。通常，玫瑰、康乃馨和百合是受欢迎的选择，因为它们既美观又有象征意义。如果你告诉我你姐姐的一些喜好，我可以给你一些建议。\n\n例如，如果她喜欢传统的花束，你可以考虑以下几种组合：\n- 红玫瑰：代表爱情和热情，适合表达对她的爱。\n- 粉玫瑰：代表温柔、优雅和感激之情。\n- 白百合：象征纯洁和尊敬，适合表达对她的敬爱。\n\n如果她喜欢特别一点的花束，你也可以考虑以下几种：\n- 向日葵：代表阳光和积极，寓意着她给你生活带来的快乐。\n- 紫罗兰：代表忠诚和深沉的爱。\n- 满天星：代表永恒，可以搭配其他花种，增加花束的丰富性。\n\n此外，你还可以考虑搭配一些绿叶植物，使花束看起来更生动、自然。在挑选花束时，记得询问花店能否提供贺卡服务，你可以写上一些祝福语，使礼物更具有个性。\n\n需要我帮你查找附近的花店或者提供更多建议吗？'

AI: 生日花束的选择取决于你姐姐的个人喜好，比如她喜欢的颜色、花种和香味。通常，玫瑰、康乃馨和百合是受欢迎的选择，因为它们既美观又有象征意义。如果你告诉我你姐姐的一些喜好，我可以给你一些建议。

例如，如果她喜欢传统的花束，你可以考虑以下几种组合：
- 红玫瑰：代表爱情和热情，适合表达对她的爱。
- 粉玫瑰：代表温柔、优雅和感激之情。
- 白百合：象征纯洁和尊敬，适合表达对她的敬爱。

如果她喜欢特别一点的花束，你也可以考虑以下几种：
- 向日葵：代表阳光和积极，寓意着她给你生活带来的快乐。
- 紫罗兰：代表忠诚和深沉的爱。
- 满天星：代表永恒，可以搭配其他花种，增加花束的丰富性。

此外，你还可以考虑搭配一些绿叶植物，使花束看起来更生动、自然。在挑选花束时，记得询问花店能否提供贺卡服务，你可以写上一些祝福语，使礼物更具有个性。

需要我帮你查找附近的花店或者提供更多建议吗？

In [14]:
conversation_with_summary.predict(input="\n她喜欢粉色玫瑰，颜色是粉色的")



> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
System: The conversation starts with the human seeking advice on maintaining a healthy lifestyle, to which the AI provides detailed recommendations. This is followed by the human asking about effective learning methods, and the AI offers a comprehensive guide. Subsequently, the human requests assistance in choosing a birthday flower bouquet for their sister. The AI suggests various flower types based on the sister's preferences, including traditional and unique options, and offers additional help such as finding nearby florists or providing further advice. 

New summary:
The AI provides health and learning advice to the human, and also assists in selecting a birthday f

"AI: That's a wonderful choice! Pink roses are a perfect symbol of gratitude, appreciation, and admiration, which makes them an ideal gift for your sister's birthday. Given her preference for pink, you could consider a few options:\n\n1. A classic bouquet of pink roses: This would be a elegant and timeless choice. You could request a mix of light and dark pink roses for a vibrant and colorful arrangement.\n\n2. A mixed bouquet with pink roses as the centerpiece: You could pair the pink roses with other flowers like white lilies or baby's breath to create a full and stunning bouquet.\n\n3. An unique arrangement: If your sister enjoys unique things, you might consider an arrangement that includes pink roses along with less common flowers, such as pink proteas, snapdragons, or carnations.\n\nIf you need help finding a florist, I can assist you with that. Would you like me to find florists near you that offer pink rose bouquets, or do you need more suggestions on the types of flowers or ar

艾：这是个不错的选择！粉色玫瑰是感激、欣赏和钦佩的完美象征，这使它们成为你姐姐生日的理想礼物。考虑到她喜欢粉色，你可以考虑几个选择：

1.一束经典的粉色玫瑰：这将是一个优雅而永恒的选择。你可以要求将浅粉色和深粉色的玫瑰混合在一起，使其充满活力和色彩。

2.以粉色玫瑰为核心的混合花束：你可以将粉色玫瑰与其他花朵（如白百合或婴儿呼吸）搭配，打造出一束饱满而迷人的花束。

3.独特的布置：如果你姐姐喜欢独特的东西，你可以考虑一种包括粉色玫瑰和不太常见的花的布置，如粉色蛋白酶、金鱼草或康乃馨。

如果你需要帮忙找花店，我可以帮你。你想让我在你附近找到提供粉色玫瑰花束的花店吗？或者你需要更多关于花的类型或布置的建议吗

In [15]:
conversation_with_summary.predict(input="\n我又来了，还记得我昨天为什么要来买花吗？")



> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
System: The AI offers comprehensive advice on health and learning to the human. Additionally, it provides tailored suggestions for a birthday flower bouquet, focusing on the sister's preference for pink roses by offering classic, mixed, and unique arrangement options. The AI also offers further assistance in locating florists or providing more flower arrangement ideas.
Human: 
我又来了，还记得我昨天为什么要来买花吗？
AI:

> Finished chain.


'AI: 当然记得，昨天您提到您的妹妹喜欢粉色的玫瑰，她的生日即将到来，您想要为她挑选一个特别的鲜花花束。如果您需要更多关于花束的建议或者想要知道哪里可以找到好的花店，我随时乐意帮助您。您今天有考虑好选择哪一种类型的花束吗？经典款、混合款，还是独特款？或者需要我再为您介绍一些其他的花艺设计理念？'

AI: 当然记得，昨天您提到您的妹妹喜欢粉色的玫瑰，她的生日即将到来，您想要为她挑选一个特别的鲜花花束。如果您需要更多关于花束的建议或者想要知道哪里可以找到好的花店，我随时乐意帮助您。您今天有考虑好选择哪一种类型的花束吗？经典款、混合款，还是独特款？或者需要我再为您介绍一些其他的花艺设计理念？